ext/systrace/systrace.c

```diff
@@ -0,0 +1,337 @@
+                 {
+                     rec->callback( x86cpu, false, rec->cb_args, syscall_ctx );
+                 }               
+
+            }
+        }
+    }
+
+    if(!is_enter){
+        free_context(syscall_ctx);
+    }
+}
+
+// Public API for add system call tracer 
+extern int systrace_add( const char* label, target_ulong target_cr3, int target_syscall_number, bool is_entry, systrace_cb callback, void* cb_args ) {
+    int new_handle;
+    systrace_record *rec;
+
+    // get empty slot for new tracer
+    for( new_handle = 0; new_handle < MAX_NM_SYSTRACE; new_handle++ )
+        if( sys_ctx->records[new_handle] == NULL )
+            break;
+
+    if( new_handle == MAX_NM_SYSTRACE )
+        FAIL( SYSTRACE_ERR_FULL_TRACE );
+
+    // alloc new record
+    rec = new_record( new_handle, label, target_cr3, target_syscall_number, is_entry, callback, cb_args );
+    if( rec == NULL )
+        FAIL( SYSTRACE_ERR_CREATE_FAIL );
+
+    pthread_rwlock_wrlock( &sys_ctx->rwlock );
+
+    // insert new record into list and occupy a table slot
+    LL_APPEND( sys_ctx->ll_records, rec );
+    sys_ctx->records[new_handle] = rec;
+
+    pthread_rwlock_unlock( &sys_ctx->rwlock );
+
+
+    return new_handle;
+}
+
+// public API for deleting system call tracer
+extern int systrace_delete( int systrace_handle ) {
+    systrace_record *rec;
+    systrace_record *tmp;
+    systrace_record *target_rec;
+    
+    // get corresponding record of given handle
+    target_rec = sys_ctx->records[systrace_handle];
+
+    if( target_rec == NULL ) {
+        FAIL( SYSTRACE_ERR_INVALID_HANDLE );
+    }
+
+    // remove record from list, freeing the record and its content
+    LL_FOREACH_SAFE( sys_ctx->ll_records, rec, tmp ) {
+        if( rec != target_rec )
+            continue;
+
+        pthread_rwlock_wrlock( &sys_ctx->rwlock );
+
+        LL_DELETE( sys_ctx->ll_records, rec );
+
+        pthread_rwlock_unlock( &sys_ctx->rwlock );
+
+        sys_ctx->records[systrace_handle] = NULL;
+        free( rec );
+        SUCCEED();
+    }
+
+    // record not in list
+    FAIL( SYSTRACE_ERR_INTERNAL_BROKEN );
+}
+
+// public API for listing all system call tracer
+extern int systrace_list(void) {
+    systrace_record *rec;
+    systrace_record *tmp;
+ 
+    printf("List all systrace\n");
+    LL_FOREACH_SAFE( sys_ctx->ll_records, rec, tmp ) {
+        printf("\tID: %d, LABEL: %s, SYSNUM: %d, CR3: %016lx, CB: %p\n", 
+              rec->id, rec->label, rec->syscall_number, rec->cr3, rec->callback);
+    }
+    SUCCEED();
+}
+
+
+// private api for implementation
+extern void systrace_on_syscall( CPUX86State *env ) {
+    return event_callret( env, true );
+}
+
+// private api for implementation
+extern void systrace_on_sysret( CPUX86State *env ) {
+    return event_callret( env, false );
+}
+
+// callback of systrace for dumping general syscall info
+void cb_log_syscall_info( X86CPU *x86cpu, bool is_invoke, void *args, const syscall_context* syscall_info ) {
```

Comment:
Similar with previous one, this fun is the demo function that print syscall num/arg on the screen. 
But the monitor is not directly pass to this function, it take time to replace printf with monitor_printf.